src/cuda_runtime_library_manager.cc

```diff
@@ -0,0 +1,117 @@
+// Copyright 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "cuda_runtime_library_manager.h"
+
+namespace triton::perfanalyzer {
+
+CUDARuntimeLibraryManager::CUDARuntimeLibraryManager()
+{
+  handle_ = dlopen("libcudart.so", RTLD_LAZY);
```

Comment:
Hmmm... would you mind trying with `RTLD_NOW` instead? `RTLD_LAZY` can be dangerous with third party libraries we don't directly maintain, even though, well, cuda is technically Nvidia's :) Sorry for the extra work. If `RTLD_NOW` works, I'd rather use it instead. The difference is that if `libcudart.so` has other transitive dependencies, they'll be all resolved immediately, and the call will fail at that time if they're not found, instead of silently failing later.

Reply:
`LoadFunctions()` which happens directly after this, calls dlsym, and checks for null returns--isn't that good enough? Perhaps even better since it doesn't load unnecessary symbols?

Reply:
The issue is it would just fail silently *at usage time* and crash, if there was, in fact, a problem loading dependencies, which is kind of counter intuitive from the nicer error message you have already in place, as it'd be completely side stepped.

Reply:
Wait you're saying dlsym can return a non-null pointer and it could crash (silently? what does a silent crash look like) when the pointer actually gets used?

Reply:
So, the full context:

`dlopen` with LAZY will always work if the file simply exists and is a valid file. Any of its own dependencies which aren't there will not be loaded, and it won't trigger an error, assuming they may be loaded and resolved later with another `dlopen`. Then, `dlsym` will only return pointers to functions, if they exist in the lookup table.

What this means is that any unresolved import inside of the library itself will just be a null pointer, as the LAZY mode only fills them in when they are loaded by other means, and the crash will be a typical uncaught segfault due to trying to access a null pointer when running code from within the library, as there is no mechanism at this point to tell if the library might be missing imports. The goal of the LAZY mechanism is to resolve circular dependencies, or solve situations where libraries are loaded on demand, like being streamed from a non-filesystem network location, but it results in a fragile environment.

On the countrary, `dlopen` with NOW means that any of the import inside of the loaded library MUST be resolved at open time, and it'll return an error if that's not the case, guaranteeing no dangling null pointer in the loaded library. It means that all of the transitives dependencies must be present already in memory, or be loadable from the filesystem.

Reply:
Done